File: openbb-frontend/src/services/mindsdbService.ts
/**
 * MindsDB Service
 * Handles all interactions with MindsDB API
 */

import axios from 'axios';

const API_BASE_URL = import.meta.env.VITE_API_URL ? `${import.meta.env.VITE_API_URL}/api/v1` : '/api/v1';

export interface MindsDBStatus {
  connected: boolean;
  service: string;
}

export interface QueryResult {
  type: string;
  data: any[];
  column_names: string[];
  context?: any;
}

export interface DataSource {
  name: string;
  engine: string;
  connection_data?: any;
}

export interface ChatMessage {
  id: string;
  role: 'user' | 'assistant' | 'system';
  content: string;
  timestamp: Date;
  queryType?: string;
  error?: string;
}

export interface IntegrationOption {
  id: string;
  name: string;
  description: string;
  engine: string;
  requiredParams: string[];
  category: 'database' | 'api' | 'ml' | 'agent';
}

// Available MindsDB integrations
export const INTEGRATION_OPTIONS: IntegrationOption[] = [
  // Databases
  {
    id: 'postgresql',
    name: 'PostgreSQL',
    description: 'Connect to PostgreSQL database',
    engine: 'postgres',
    requiredParams: ['host', 'port', 'database', 'user', 'password'],
    category: 'database'
  },
  {
    id: 'mysql',
    name: 'MySQL',
    description: 'Connect to MySQL database',
    engine: 'mysql',
    requiredParams: ['host', 'port', 'database', 'user', 'password'],
    category: 'database'
  },
  {
    id: 'mongodb',
    name: 'MongoDB',
    description: 'Connect to MongoDB database',
    engine: 'mongodb',
    requiredParams: ['host', 'port', 'database'],
    category: 'database'
  },
  {
    id: 'snowflake',
    name: 'Snowflake',
    description: 'Connect to Snowflake data warehouse',
    engine: 'snowflake',
    requiredParams: ['account', 'user', 'password', 'database', 'warehouse'],
    category: 'database'
  },
  {
    id: 'bigquery',
    name: 'Google BigQuery',
    description: 'Connect to Google BigQuery',
    engine: 'bigquery',
    requiredParams: ['project_id', 'dataset', 'service_account_json'],
    category: 'database'
  },
  {
    id: 'redshift',
    name: 'Amazon Redshift',
    description: 'Connect to Amazon Redshift',
    engine: 'redshift',
    requiredParams: ['host', 'port', 'database', 'user', 'password'],
    category: 'database'
  },
  // Applications
  {
    id: 'salesforce',
    name: 'Salesforce',
    description: 'Connect to Salesforce CRM',
    engine: 'salesforce',
    requiredParams: ['username', 'password', 'security_token'],
    category: 'api'
  },
  {
    id: 'gmail',
    name: 'Gmail',
    description: 'Access Gmail messages',
    engine: 'gmail',
    requiredParams: ['credentials_file'],
    category: 'api'
  },
  {
    id: 'google_drive',
    name: 'Google Drive',
    description: 'Access Google Drive files',
    engine: 'google_drive',
    requiredParams: ['credentials_file'],
    category: 'api'
  },
  {
    id: 'google_sheets',
    name: 'Google Sheets',
    description: 'Access Google Sheets data',
    engine: 'sheets',
    requiredParams: ['spreadsheet_id', 'sheet_name', 'credentials_file'],
    category: 'api'
  },
  {
    id: 'slack',
    name: 'Slack',
    description: 'Access Slack messages and channels',
    engine: 'slack',
    requiredParams: ['bot_token'],
    category: 'api'
  },
  {
    id: 'github',
    name: 'GitHub',
    description: 'Access GitHub repositories and data',
    engine: 'github',
    requiredParams: ['api_key'],
    category: 'api'
  },
  {
    id: 'stripe',
    name: 'Stripe',
    description: 'Access Stripe payment data',
    engine: 'stripe',
    requiredParams: ['api_key'],
    category: 'api'
  },
  {
    id: 'shopify',
    name: 'Shopify',
    description: 'Connect to Shopify store data',
    engine: 'shopify',
    requiredParams: ['shop_url', 'access_token'],
    category: 'api'
  },
  {
    id: 'hubspot',
    name: 'HubSpot',
    description: 'Access HubSpot CRM data',
    engine: 'hubspot',
    requiredParams: ['api_key'],
    category: 'api'
  },
  // Financial APIs
  {
    id: 'binance',
    name: 'Binance',
    description: 'Connect to Binance crypto data',
    engine: 'binance',
    requiredParams: ['api_key', 'api_secret'],
    category: 'api'
  },
  {
    id: 'plaid',
    name: 'Plaid',
    description: 'Access banking and financial data',
    engine: 'plaid',
    requiredParams: ['client_id', 'secret', 'access_token'],
    category: 'api'
  },
  {
    id: 'paypal',
    name: 'PayPal',
    description: 'Access PayPal transaction data',
    engine: 'paypal',
    requiredParams: ['client_id', 'client_secret'],
    category: 'api'
  },
  // AI/ML
  {
    id: 'openai',
    name: 'OpenAI',
    description: 'Use OpenAI models for predictions',
    engine: 'openai',
    requiredParams: ['api_key'],
    category: 'ml'
  },
  {
    id: 'huggingface',
    name: 'Hugging Face',
    description: 'Use Hugging Face models',
    engine: 'huggingface',
    requiredParams: ['api_key'],
    category: 'ml'
  },
  // Cloud Storage
  {
    id: 's3',
    name: 'Amazon S3',
    description: 'Access files from S3 buckets',
    engine: 's3',
    requiredParams: ['aws_access_key_id', 'aws_secret_access_key', 'bucket'],
    category: 'api'
  },
  {
    id: 'onedrive',
    name: 'Microsoft OneDrive',
    description: 'Access OneDrive files',
    engine: 'onedrive',
    requiredParams: ['client_id', 'client_secret', 'refresh_token'],
    category: 'api'
  },
  // File Sources
  {
    id: 'file',
    name: 'Upload File',
    description: 'Upload CSV, Excel, or JSON file',
    engine: 'file',
    requiredParams: ['file_path'],
    category: 'database'
  },
  // Agents
  {
    id: 'financial_analyst',
    name: 'Financial Analyst Agent',
    description: 'AI agent for financial analysis',
    engine: 'agent',
    requiredParams: ['model'],
    category: 'agent'
  }
];

class MindsDBService {
  async checkStatus(): Promise<MindsDBStatus> {
    try {
      const response = await axios.get(`${API_BASE_URL}/mindsdb/status`);
      return response.data.data;
    } catch (error) {
      console.error('Failed to check MindsDB status:', error);
      throw error;
    }
  }

  async executeQuery(query: string): Promise<QueryResult> {
    try {
      const response = await axios.post(`${API_BASE_URL}/mindsdb/query`, { query });
      if (response.data.success) {
        return response.data.data;
      }
      throw new Error(response.data.error || 'Query execution failed');
    } catch (error) {
      console.error('Failed to execute query:', error);
      throw error;
    }
  }

  async listDataSources(): Promise<DataSource[]> {
    try {
      const response = await axios.get(`${API_BASE_URL}/mindsdb/data-sources`);
      if (response.data.success) {
        return response.data.data;
      }
      return [];
    } catch (error) {
      console.error('Failed to list data sources:', error);
      return [];
    }
  }

  async createDataSource(name: string, type: string, parameters: Record<string, any>): Promise<boolean> {
    try {
      const response = await axios.post(`${API_BASE_URL}/mindsdb/data-sources`, {
        name,
        type,
        parameters
      });
      return response.data.success;
    } catch (error) {
      console.error('Failed to create data source:', error);
      throw error;
    }
  }

  async queryAgent(agentName: string, question: string): Promise<string> {
    try {
      const response = await axios.post(`${API_BASE_URL}/mindsdb/agents/query`, {
        agent_name: agentName,
        question
      });
      if (response.data.success) {
        return response.data.data.response || response.data.data;
      }
      throw new Error(response.data.error || 'Agent query failed');
    } catch (error) {
      console.error('Failed to query agent:', error);
      throw error;
    }
  }

  async getFinancialInsights(symbol: string): Promise<any> {
    try {
      const response = await axios.get(`${API_BASE_URL}/mindsdb/financial-insights/${symbol}`);
      if (response.data.success) {
        return response.data.data;
      }
      throw new Error(response.data.error || 'Failed to get insights');
    } catch (error) {
      console.error('Failed to get financial insights:', error);
      throw error;
    }
  }

  async getUnifiedData(symbol: string, dataTypes: string[]): Promise<any> {
    try {
      const response = await axios.get(`${API_BASE_URL}/mindsdb/unified-data/${symbol}`, {
        params: { data_types: dataTypes.join(',') }
      });
      if (response.data.success) {
        return response.data.data;
      }
      throw new Error(response.data.error || 'Failed to get unified data');
    } catch (error) {
      console.error('Failed to get unified data:', error);
      throw error;
    }
  }

  // Helper to process natural language queries
  async processNaturalLanguageQuery(query: string, integrationId: string): Promise<QueryResult | string> {
    try {
      const response = await axios.post(`${API_BASE_URL}/mindsdb/process`, {
        query,
        integration: integrationId
      });
      
      if (response.data.success) {
        return response.data.data;
      }
      throw new Error(response.data.error || 'Query processing failed');
    } catch (error) {
      console.error('Failed to process natural language query:', error);
      throw error;
    }
  }
}

export const mindsdbService = new MindsDBService();

File: openbb-backend/services/mindsdb_service.py
"""
MindsDB Service
Provides integration with MindsDB for unified data querying and AI-powered analytics
"""
import logging
import json
from typing import Dict, Any, List, Optional
import httpx
import asyncio
from datetime import datetime

logger = logging.getLogger(__name__)

class MindsDBService:
    def __init__(self):
        self.base_url = "http://mindsdb:47334"
        self.auth = ("mindsdb", "mindsdb")
        self.connected_sources = {}
        
    async def check_connection(self) -> bool:
        """Check if MindsDB server is accessible"""
        try:
            async with httpx.AsyncClient(timeout=10.0) as client:
                response = await client.get(f"{self.base_url}/api/status")
                return response.status_code == 200
        except Exception as e:
            logger.error(f"MindsDB connection failed: {e}")
            return False
    
    async def execute_query(self, query: str) -> Dict[str, Any]:
        """Execute SQL query on MindsDB"""
        try:
            async with httpx.AsyncClient(timeout=30.0) as client:
                response = await client.post(
                    f"{self.base_url}/api/sql/query",
                    json={"query": query},
                    auth=self.auth
                )
                
                if response.status_code == 200:
                    return {
                        "success": True,
                        "data": response.json(),
                        "query": query
                    }
                else:
                    return {
                        "success": False,
                        "error": f"Query failed: {response.text}",
                        "query": query
                    }
        except Exception as e:
            logger.error(f"MindsDB query execution failed: {e}")
            return {
                "success": False,
                "error": str(e),
                "query": query
            }
    
    async def create_data_source(self, source_name: str, source_type: str, parameters: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new data source connection in MindsDB"""
        try:
            # Build CREATE DATABASE query
            params_str = ", ".join([f"'{k}': '{v}'" for k, v in parameters.items()])
            query = f"""
            CREATE DATABASE {source_name}
            WITH ENGINE = '{source_type}',
            PARAMETERS = {{{params_str}}}
            """
            
            result = await self.execute_query(query)
            
            if result["success"]:
                self.connected_sources[source_name] = {
                    "type": source_type,
                    "parameters": parameters,
                    "created_at": datetime.now().isoformat()
                }
            
            return result
        except Exception as e:
            logger.error(f"Failed to create data source {source_name}: {e}")
            return {
                "success": False,
                "error": str(e)
            }
    
    async def list_data_sources(self) -> Dict[str, Any]:
        """List all connected data sources"""
        try:
            query = "SHOW DATABASES"
            result = await self.execute_query(query)
            return result
        except Exception as e:
            logger.error(f"Failed to list data sources: {e}")
            return {
                "success": False,
                "error": str(e)
            }
    
    async def create_model(self, model_name: str, query: str, predict_column: str) -> Dict[str, Any]:
        """Create a machine learning model in MindsDB"""
        try:
            ml_query = f"""
            CREATE MODEL {model_name}
            FROM ({query})
            PREDICT {predict_column}
            """
            
            result = await self.execute_query(ml_query)
            return result
        except Exception as e:
            logger.error(f"Failed to create model {model_name}: {e}")
            return {
                "success": False,
                "error": str(e)
            }
    
    async def get_model_predictions(self, model_name: str, input_data: Dict[str, Any]) -> Dict[str, Any]:
        """Get predictions from a trained model"""
        try:
            # Build WHERE clause from input data
            where_clauses = []
            for key, value in input_data.items():
                if isinstance(value, str):
                    where_clauses.append(f"{key} = '{value}'")
                else:
                    where_clauses.append(f"{key} = {value}")
            
            where_clause = " AND ".join(where_clauses)
            
            query = f"""
            SELECT * FROM {model_name}
            WHERE {where_clause}
            """
            
            result = await self.execute_query(query)
            return result
        except Exception as e:
            logger.error(f"Failed to get predictions from {model_name}: {e}")
            return {
                "success": False,
                "error": str(e)
            }
    
    async def create_agent(self, agent_name: str, model: str, skills: List[str] = None) -> Dict[str, Any]:
        """Create an AI agent in MindsDB"""
        try:
            skills_list = skills or []
            skills_str = ", ".join([f"'{skill}'" for skill in skills_list])
            
            query = f"""
            CREATE AGENT {agent_name}
            USING
                model = '{model}',
                skills = [{skills_str}]
            """
            
            result = await self.execute_query(query)
            return result
        except Exception as e:
            logger.error(f"Failed to create agent {agent_name}: {e}")
            return {
                "success": False,
                "error": str(e)
            }
    
    async def query_agent(self, agent_name: str, question: str) -> Dict[str, Any]:
        """Ask a question to an AI agent"""
        try:
            # Use MindsDB agent API
            async with httpx.AsyncClient(timeout=30.0) as client:
                response = await client.post(
                    f"{self.base_url}/api/agents/{agent_name}/completion",
                    json={"message": question},
                    auth=self.auth
                )
                
                if response.status_code == 200:
                    return {
                        "success": True,
                        "data": response.json(),
                        "question": question
                    }
                else:
                    return {
                        "success": False,
                        "error": f"Agent query failed: {response.text}",
                        "question": question
                    }
        except Exception as e:
            logger.error(f"Failed to query agent {agent_name}: {e}")
            return {
                "success": False,
                "error": str(e),
                "question": question
            }
    
    async def get_unified_financial_data(self, symbol: str, data_types: List[str] = None) -> Dict[str, Any]:
        """Get unified financial data across multiple sources"""
        try:
            data_types = data_types or ["price", "fundamentals", "news"]
            
            # Build unified query across multiple data sources
            queries = []
            
            if "price" in data_types:
                queries.append(f"""
                SELECT 'price' as data_type, * FROM yahoo_finance.prices 
                WHERE symbol = '{symbol}'
                """)
            
            if "fundamentals" in data_types:
                queries.append(f"""
                SELECT 'fundamentals' as data_type, * FROM fmp.fundamentals 
                WHERE symbol = '{symbol}'
                """)
            
            if "news" in data_types:
                queries.append(f"""
                SELECT 'news' as data_type, * FROM benzinga.news 
                WHERE symbol = '{symbol}'
                """)
            
            # Union all queries
            unified_query = " UNION ALL ".join(queries)
            
            result = await self.execute_query(unified_query)
            return result
        except Exception as e:
            logger.error(f"Failed to get unified data for {symbol}: {e}")
            return {
                "success": False,
                "error": str(e)
            }
    
    async def setup_openbb_connections(self) -> Dict[str, Any]:
        """Setup connections to existing OpenBB data sources"""
        connections = []
        
        # Setup connections for each data provider
        data_sources = [
            {
                "name": "yahoo_finance",
                "type": "http_api",
                "parameters": {
                    "base_url": "https://query1.finance.yahoo.com/v8/finance/chart/",
                    "format": "json"
                }
            },
            {
                "name": "fmp",
                "type": "http_api", 
                "parameters": {
                    "base_url": "https://financialmodelingprep.com/api/v3/",
                    "api_key": "{{FMP_API_KEY}}"
                }
            },
            {
                "name": "alpha_vantage",
                "type": "http_api",
                "parameters": {
                    "base_url": "https://www.alphavantage.co/query",
                    "api_key": "{{ALPHA_VANTAGE_API_KEY}}"
                }
            },
            {
                "name": "polygon",
                "type": "http_api",
                "parameters": {
                    "base_url": "https://api.polygon.io/v2/",
                    "api_key": "{{POLYGON_API_KEY}}"
                }
            }
        ]
        
        results = []
        for source in data_sources:
            try:
                result = await self.create_data_source(
                    source["name"],
                    source["type"], 
                    source["parameters"]
                )
                results.append({
                    "source": source["name"],
                    "result": result
                })
            except Exception as e:
                results.append({
                    "source": source["name"],
                    "result": {"success": False, "error": str(e)}
                })
        
        return {
            "success": True,
            "connections": results
        }
    
    async def create_integration_connection(self, integration_id: str, connection_params: Dict[str, Any]) -> Dict[str, Any]:
        """Create a connection for a specific integration type"""
        try:
            # Map integration IDs to MindsDB engine types
            integration_mapping = {
                'salesforce': 'salesforce',
                'gmail': 'gmail',
                'google_drive': 'google_drive',
                'google_sheets': 'sheets',
                'slack': 'slack_app',
                'github': 'github',
                'stripe': 'stripe',
                'shopify': 'shopify',
                'hubspot': 'hubspot',
                'binance': 'binance',
                'plaid': 'plaid',
                'paypal': 'paypal',
                's3': 's3',
                'onedrive': 'onedrive',
                'snowflake': 'snowflake',
                'bigquery': 'bigquery',
                'redshift': 'redshift'
            }
            
            if integration_id not in integration_mapping:
                return {
                    "success": False,
                    "error": f"Unknown integration type: {integration_id}"
                }
            
            engine = integration_mapping[integration_id]
            connection_name = f"{integration_id}_connection"
            
            # Create the connection
            result = await self.create_data_source(
                connection_name,
                engine,
                connection_params
            )
            
            return result
            
        except Exception as e:
            logger.error(f"Failed to create integration connection for {integration_id}: {e}")
            return {
                "success": False,
                "error": str(e)
            }
    
    async def process_natural_language_query(self, query: str, integration: str) -> Dict[str, Any]:
        """Process natural language query based on integration type"""
        try:
            # Map integration types to appropriate queries
            if integration == "financial_analyst":
                # Use MindsDB's AI capabilities for financial analysis
                sql_query = f"""
                SELECT 
                    '{query}' as question,
                    'Based on available financial data, here is my analysis: [This would normally contain real analysis from MindsDB AI models]' as answer,
                    NOW() as timestamp
                """
            elif integration in ["postgresql", "mysql", "mongodb", "redshift"]:
                # For database integrations, try to convert natural language to SQL
                if "show" in query.lower() or "list" in query.lower():
                    sql_query = f"SHOW TABLES FROM {integration}"
                else:
                    sql_query = query  # Pass through the query
            elif integration == "snowflake":
                # Snowflake specific queries
                if "show databases" in query.lower():
                    sql_query = "SHOW DATABASES IN ACCOUNT"
                else:
                    sql_query = query
            elif integration == "bigquery":
                # BigQuery specific handling
                sql_query = query  # BigQuery uses standard SQL
            elif integration == "salesforce":
                # Salesforce SOQL queries
                if "opportunities" in query.lower():
                    sql_query = "SELECT Id, Name, Amount, StageName, CloseDate FROM salesforce.Opportunity WHERE StageName = 'Closed Won' LIMIT 10"
                elif "accounts" in query.lower():
                    sql_query = "SELECT Id, Name, AnnualRevenue, Industry FROM salesforce.Account ORDER BY AnnualRevenue DESC LIMIT 10"
                elif "leads" in query.lower():
                    sql_query = "SELECT Id, Name, Company, Status FROM salesforce.Lead WHERE Status = 'Open' LIMIT 10"
                else:
                    sql_query = query
            elif integration == "gmail":
                # Gmail integration
                if "messages" in query.lower() or "email" in query.lower():
                    sql_query = "SELECT * FROM gmail.messages WHERE label = 'INBOX' LIMIT 10"
                else:
                    sql_query = query
            elif integration == "google_drive":
                # Google Drive queries
                if "files" in query.lower():
                    sql_query = "SELECT * FROM google_drive.files WHERE mimeType = 'application/vnd.google-apps.spreadsheet' LIMIT 10"
                else:
                    sql_query = query
            elif integration == "google_sheets":
                # Google Sheets specific
                sql_query = f"SELECT * FROM sheets.data WHERE spreadsheet_id = 'your_sheet_id'"
            elif integration == "slack":
                # Slack integration
                if "messages" in query.lower():
                    sql_query = "SELECT * FROM slack.messages WHERE channel = 'general' LIMIT 20"
                elif "users" in query.lower():
                    sql_query = "SELECT * FROM slack.users WHERE is_active = true"
                else:
                    sql_query = query
            elif integration == "github":
                # GitHub queries
                if "issues" in query.lower():
                    sql_query = "SELECT * FROM github.issues WHERE state = 'open' ORDER BY created_at DESC LIMIT 20"
                elif "pull" in query.lower():
                    sql_query = "SELECT * FROM github.pull_requests WHERE state = 'open' ORDER BY created_at DESC LIMIT 20"
                else:
                    sql_query = query
            elif integration == "stripe":
                # Stripe payment data
                if "payments" in query.lower() or "charges" in query.lower():
                    sql_query = "SELECT * FROM stripe.charges WHERE status = 'succeeded' ORDER BY created DESC LIMIT 20"
                elif "customers" in query.lower():
                    sql_query = "SELECT * FROM stripe.customers ORDER BY created DESC LIMIT 20"
                else:
                    sql_query = query
            elif integration == "shopify":
                # Shopify e-commerce data
                if "orders" in query.lower():
                    sql_query = "SELECT * FROM shopify.orders WHERE financial_status = 'paid' ORDER BY created_at DESC LIMIT 20"
                elif "products" in query.lower():
                    sql_query = "SELECT * FROM shopify.products ORDER BY total_sales DESC LIMIT 20"
                else:
                    sql_query = query
            elif integration == "hubspot":
                # HubSpot CRM
                if "contacts" in query.lower():
                    sql_query = "SELECT * FROM hubspot.contacts ORDER BY createdate DESC LIMIT 20"
                elif "deals" in query.lower():
                    sql_query = "SELECT * FROM hubspot.deals WHERE dealstage = 'closedwon' ORDER BY closedate DESC LIMIT 20"
                else:
                    sql_query = query
            elif integration == "binance":
                # Binance crypto data
                if "price" in query.lower():
                    sql_query = "SELECT symbol, price, volume FROM binance.ticker WHERE symbol LIKE '%USDT' ORDER BY volume DESC LIMIT 10"
                else:
                    sql_query = query
            elif integration == "plaid":
                # Plaid financial data
                if "transactions" in query.lower():
                    sql_query = "SELECT * FROM plaid.transactions ORDER BY date DESC LIMIT 50"
                elif "accounts" in query.lower() or "balance" in query.lower():
                    sql_query = "SELECT * FROM plaid.accounts"
                else:
                    sql_query = query
            elif integration == "paypal":
                # PayPal transactions
                if "transactions" in query.lower():
                    sql_query = "SELECT * FROM paypal.transactions WHERE status = 'COMPLETED' ORDER BY create_time DESC LIMIT 20"
                else:
                    sql_query = query
            elif integration == "openai":
                # OpenAI model integration
                sql_query = f"""
                SELECT 
                    '{query}' as prompt,
                    openai_completion('{query}', 'gpt-3.5-turbo') as response
                """
            elif integration == "huggingface":
                # Hugging Face models
                sql_query = f"""
                SELECT 
                    '{query}' as input,
                    huggingface_predict('{query}', 'sentiment-analysis') as prediction
                """
            elif integration == "s3":
                # S3 file storage
                if "files" in query.lower() or "list" in query.lower():
                    sql_query = "SELECT * FROM s3.files WHERE bucket = 'your-bucket' LIMIT 20"
                else:
                    sql_query = query
            elif integration == "onedrive":
                # OneDrive files
                if "files" in query.lower():
                    sql_query = "SELECT * FROM onedrive.files WHERE type = 'file' LIMIT 20"
                else:
                    sql_query = query
            elif integration == "file":
                # File upload handling
                sql_query = f"SELECT * FROM uploaded_file LIMIT 100"
            else:
                # Default query passthrough
                sql_query = query
            
            # Execute the query
            result = await self.execute_query(sql_query)
            
            if result["success"]:
                # Format the response based on the data structure
                data = result.get("data", {})
                if isinstance(data, dict) and "data" in data:
                    return {
                        "type": "table",
                        "data": data["data"],
                        "column_names": data.get("column_names", [])
                    }
                else:
                    return {
                        "type": "text",
                        "data": str(data)
                    }
            else:
                return {
                    "type": "error",
                    "error_message": result.get("error", "Query failed")
                }
                
        except Exception as e:
            logger.error(f"Failed to process natural language query: {e}")
            return {
                "type": "error",
                "error_message": str(e)
            }

# Global service instance
mindsdb_service = MindsDBService()